django/utils/decorators.py

```diff
@@ -17,13 +17,29 @@ def __get__(self, instance, owner):
         return super(classonlymethod, self).__get__(instance, owner)
 
 
-def method_decorator(decorator):
+def method_decorator(decorator, name=""):
     """
     Converts a function decorator into a method decorator
     """
     # 'func' is a function at the time it is passed to _dec, but will eventually
     # be a method of the class it is defined on.
-    def _dec(func):
+    # 'name' is required when a class is decorated instead of a method,
+    # and is the name of the method to be decorated, e.g. "dispatch".
+    def _dec(obj):
+        is_class = isinstance(obj, type)
+        if is_class:
+            if name and hasattr(obj, name):
+                func = getattr(obj, name)
```

Comment:
Done. If `func` is not callable, a `TypeError` with an informative message will be raised. I've also added a test for this case.